Story 3 view buildings on map

mmc03-ucb opened the pull request:
Implemented MapView and MapViewModel

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+class MapViewModel: ObservableObject {
+  @Published var buildings: [Building] = []
+
+  private let buildingService = BuildingService()
```
This BuildingService is an implicit dependency since it is not explicitly injected into the class (either in the initialiser or a method). We want to minimise the number of implicit dependencies as much as possible as it is difficult to control them in unit tests (e.g. injecting a mock). I recommend injecting it in the initialiser.

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+import Buildings
+
+class MapViewModel: ObservableObject {
```
Since we have iOS 17 as a minimum deployment target we can use the [@Observable macro](https://developer.apple.com/documentation/observation/observable())

This API cleans up the code since it removes the need for @Published properties and lets us easily have nested observables.

avdn-dev replied:
https://developer.apple.com/documentation/swiftui/migrating-from-the-observable-object-protocol-to-the-observable-macro

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+  private let buildingService = BuildingService()
+
+  init() {
+    loadBuildings()
```
Since this loadBuildings call will result in a potentially slow network request, we want to avoid blocking any other initialisers or important work. Generally initialisers should have the bare minimum required to initialise the unit. In this case it's ok if the buildings are initially empty (we can display some loading state).

avdn-dev reviewed (changes requested):
Great start! Just a couple comments to resolve regarding modern API usage and best practice.

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+import Buildings
+
+class MapViewModel: ObservableObject {
```
As we likely won't subclass this view model, marking it as final will help optimise the code by allowing for Swift to statically dispatch methods and properties instead of dynamically dispatching them at run time. You can find more details here: https://developer.apple.com/swift/blog/?id=27

mmc03-ucb commented:
Replaced ObservableObject and @Published with @Observable macro.

Injected BuildingService through the initializer to remove implicit dependency.

Marked MapViewModel as final for performance optimization.

Removed loadBuildings() from the initializer to avoid blocking during init.

Made BuildingService public to avoid out of scope error

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingService.swift:
```diff
@@ -6,9 +6,11 @@
-
-class BuildingService {
-  func getBuildings() -> [Building] {
+// Made this public to avoid out of scope error for MapViewModel
```
[NIT] Generally we don't need to comment on why something has been given a specific scope. To promote encapsulation we opt for more restrictive access modifiers by default, so if something is set to public then it must be because it has to be visible from another package.

avdn-dev commented on ios/Buildings/Sources/Buildings/BuildingService.swift:
```diff
@@ -6,9 +6,11 @@
-class BuildingService {
-  func getBuildings() -> [Building] {
+// Made this public to avoid out of scope error for MapViewModel
+public class BuildingService {
```
I forgot to comment that this one should also be final

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+import Foundation 
+import Observation 
+
+// The @Observable macro automatically makes this class observable for UI updates.
```
[NIT] This comment isn't needed as all view models will have this macro

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+// The @Observable macro automatically makes this class observable for UI updates.
+@Observable
+final class MapViewModel {
+    // The UI will automatically update when this array changes because of @Observable.
```
[NIT] While comments are generally good, having too many runs the risk of more stale comments + cluttering the code with over-explanation. A rule of thumb with comments is they should explain why code is written as is (in the context of the codebase) instead of explaining what it does. In this case it's a comment that explains how the observation framework works, which is very general to iOS development and not unique to our codebase.

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+    // The UI will automatically update when this array changes because of @Observable.
+    var buildings: [Building] = []
+    
+    // Reference to a service that provides building data.
```
[NIT] This documentation can be added directly to the BuildingService itself

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+    // Reference to a service that provides building data.
+    private let buildingService: BuildingService
+
+    // Custom BuildingService (for example, a mock for testing), or use the default one for normal app usage.
```
[NIT] Similar to the comment above it is redundant as it is a common pattern

avdn-dev commented on ios/Freerooms/MapViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+        self.buildingService = buildingService 
+    }
+
+    // Loads building data from the service and updates the buildings array.
```
[NIT] This is a repeat of exactly how the function is written - while it might be over explaining that can be removed now, you could update it instead to practice writing Swift documentation comments like so: https://developer.apple.com/documentation/xcode/writing-symbol-documentation-in-your-source-files

avdn-dev reviewed (changes requested):
Mostly commenting issues to be resolved now

mmc03-ucb commented:
Removed redundant comments and made BuildingService final
